Added get multiple api for badger

harshil-goel opened the pull request.

darkcoderrises committed:
Added get multiple api for badger

darkcoderrises committed:
updated getBatch

mangalaman93 commented on db.go:
```diff
@@ -747,7 +749,54 @@ func (db *DB) getMemTables() ([]*memTable, func()) {
+		if keysRead[j] {
+			continue
+		}
+		version := y.ParseTs(key)
```
this can be an argument to the function

mangalaman93 reviewed (changes requested):
1. move the common functions out and use them in getBatch and get
2. Avoid allocating vss and seeks slices
3. define a function for converting ValueStruct to Item type
4. Consider taking the []*Item as an argument instead of constructing it inside GetBatch

darkcoderrises committed:
fix some more comments

shiva-istari committed:
fix: repair getBatch signature drift from PR tip

The PR's last commit threaded 'version uint64' through
db.getBatch -> lc.getBatch but left two call sites inconsistent, so the
PR tip never compiled:
- db.get passed 2 args to the now-3-arg db.getBatch; pass the version
  parsed from the ts-encoded key, matching the non-batch path below.
- lc.getBatch passed version down to levelHandler.getBatch, whose
  signature never took it (and doesn't need it - version filtering
  happens in lc.getBatch); drop the argument.

Build repair only: the batching behavior (and its bugs) is unchanged.

shiva-istari committed:
test: pin Txn.GetBatch semantics with definite insert->assert tests

Fifteen tests establishing the contract that GetBatch(keys)[i] must observe
exactly what Get(keys[i]) observes at the same read timestamp: committed
versions, the transaction's own pending writes and deletes, nil items for
absent keys, value-log values, unsorted keys across memtable and LSM, L0
version placement across reused iterator sets, reads at exactly the commit
timestamp, SSI conflict tracking, TTL expiry mid-batch, banned namespaces,
and snapshot stability under concurrent writes, flushes and compactions.

At this commit several of these tests FAIL (some panic): they document the
GetBatch bugs present in the implementation so far, and are turned green by
the following commit.

shiva-istari committed:
fix: make Txn.GetBatch a correct batched equivalent of Txn.Get

Repairs the batched read path end to end; turns the previous commit's
tests green:

- pendingWrites: nil-deref on non-update txns, read-your-own-delete
  resurrecting committed versions, and fulfilled slots clobbered by their
  unused LSM results.
- Absent keys returned impostor empty Items; they now return nil items,
  the batched analogue of ErrKeyNotFound.
- L0: the reusable per-batch iterator set bloom-skipped overlapping L0
  tables that lacked the first key, hiding newer versions of later keys
  (MVCC stale read). L0 tables are no longer bloom-skipped when building
  a reusable set; levels >= 1 are non-overlapping and still skip.
- Table ref leak: iterator-set rebuilds overwrote the decr closure
  without invoking it, pinning table references.
- Found/not-found discriminator now uses Version==0 instead of
  len(Value)==0: tombstones and empty values are answers, not misses.
- done/keysRead aliasing: the slice marking "answered from pendingWrites"
  was also handed to db.getBatch as its exact-version-hit scratch, so a
  version committed exactly at readTs was skipped during item assembly
  and read back as absent. Pending fulfillment now has its own slice.

shiva-istari committed:
chore: check iterator Close error returns in getBatch paths

Appeases errcheck: assign the error returns of table iterator Close
calls in levelHandler.getBatch's close_iters helper and in
levelHandler.get to blank, matching the pattern used elsewhere in the
repo. No behavior change.
